test/modifiers/batch.test.js

```diff
@@ -0,0 +1,284 @@
+var expect = require('chai').expect;
+var _ = require('lodash');
+var through = require('through2');
+
+var getReadableStream = require('../_utilities/getReadableStream.js');
+var runBasicStreamTests = require('../_utilities/runBasicStreamTests.js');
+
+var batch = require('../../').batch;
+
+function buff(str) {
+  return new Buffer(str);
+}
+
+describe('[throttle]', function () {
+  var data = ['item1', new Buffer('item2'), 'item3', 'item4'];
+  var objData = [true, false, [1, 2, 3], 'string', 0, '11', 95.23, { obj: true }, _.noop];
+
+  function basicTests(batchOptions, expected, objExpected) {
+    runBasicStreamTests(data, objData, function (stream, objectMode, done) {
+      var actual = [];
+
+      stream
+        .pipe(batch(batchOptions))
+        .on('data', function (chunk) {
+          actual.push(chunk);
+        })
+        .on('error', done)
+        .on('end', function () {
+          if (objectMode) {
+            expect(actual).to.deep.equal(objExpected);
+          } else {
+            expect(actual).to.deep.equal(expected);
+          }
+
+          done();
+        });
+    });
+
+    it('writes all chunks as an array', function (done) {
+      var input = getReadableStream([1, 2, 3], {
+        objectMode: true
+      });
+
+      var chunksRead = false;
+
+      input
+        .pipe(batch(batchOptions))
+        .on('data', function (chunk) {
+          chunksRead = true;
+          expect(chunk).to.be.an('array');
+        })
+        .on('error', done)
+        .on('end', function () {
+          expect(chunksRead).to.equal(true);
+          done();
+        });
+    });
+
+    it('successfully ends if there is no data written', function (done) {
+      var input = getReadableStream([], {
+        objectMode: true
+      });
+
+      var chunksRead = false;
+
+      input
+        .pipe(batch(batchOptions))
+        .on('data', function (chunk) {
+          chunksRead = true;
+          expect(chunk).to.be.an('array');
+        })
+        .on('error', done)
+        .on('end', function () {
+          expect(chunksRead).to.equal(false);
+          done();
+        });
+    });
+
+  }
+
+  describe('when "time" is defined in options', function () {
+    var expected = [[buff('item1')], [buff('item2'), buff('item3'), buff('item4')]];
+    var objExpected = [
+      objData.slice(0, 1),
+      objData.slice(1)
+    ];
+
+    basicTests({ time: 5 }, expected, objExpected);
```

Comment:
I am not totally sure about this one. The stream helpers you have in the basic tests do all the writes synchronously. Further, the expectations are basically that there will be two writes: the first with only the first item and the second with all the rest of the items... effectively only testing the `leading` aspect of `batch`. The thing it really does do is make sure that `batch` supports the `ReadableStream` and `DuplexStream` interfaces, and not much else. (Is this why I felt safe using `through` for the rest of the tests?)

I am not totally sure how I would integrate fake time into this, mostly because the streams themselves are synchronous and the tests are actually in the `runBasicTests` helper and are used throughout all of `node-stream` to run tests. I would have to think about this one. I don't see these tests being flaky like the other one though.